source/resource_estimator/src/counts/memory_compute.rs

```diff
@@ -0,0 +1,339 @@
+    removed: usize,
+    // Maximum size reached at any point in time.
+    max_size: usize,
+}
+
+impl<K: Eq + Hash + Clone> LruPQ<K> {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            map: FxHashSet::with_capacity_and_hasher(capacity, Default::default()),
+            nodes: VecDeque::with_capacity(capacity),
+            capacity,
+            inserted_new: 0,
+            removed: 0,
+            max_size: 0,
+        }
+    }
+
+    pub fn inserted_new_count(&self) -> usize {
+        self.inserted_new
+    }
+
+    pub fn removed_count(&self) -> usize {
+        self.removed
+    }
+
+    pub fn max_size(&self) -> usize {
+        self.max_size
+    }
+
+    fn contains(&self, key: &K) -> bool {
+        self.map.contains(key)
+    }
+
+    /// Insert multiple keys ensuring they are all present afterwards. If more
+    /// unique new keys are provided than capacity, only the most recently
+    /// processed up to `capacity` will remain.
+    pub fn insert_all<I: IntoIterator<Item = K>>(&mut self, keys: I) {
+        if self.capacity == 0 {
+            return;
+        }
+        // Collect unique keys from input preserving order of first occurrence.
+        let mut seen_input = FxHashSet::default();
+        let mut ordered: Vec<K> = Vec::new();
+        for k in keys {
+            if seen_input.insert(k.clone()) {
+                ordered.push(k);
+            }
+        }
+        debug_assert!(
+            ordered.len() <= self.capacity,
+            "More keys than capacity in LruPQ::insert_all"
+        );
+
+        // Process each key in order; we evict as we go and since new elements
+        // are moved front they will be retained if we exceed capacity.
+        for k in ordered {
+            if self.contains(&k) {
+                // Just update recency by moving element to front of deque
+                if let Some(value) = self
+                    .nodes
+                    .iter()
+                    .position(|n| n == &k)
+                    .and_then(|i| self.nodes.remove(i))
+                {
+                    self.nodes.push_front(value);
+                }
+            } else {
+                // Evict if at capacity
+                if self.map.len() == self.capacity {
+                    if let Some(key) = self.nodes.pop_back() {
+                        self.map.remove(&key);
+                        self.removed += 1;
+                    }
+                }
+                self.map.insert(k.clone());
+                self.nodes.push_front(k);
+                self.inserted_new += 1;
+            }
+        }
+
+        if self.map.len() > self.max_size {
+            self.max_size = self.map.len();
+        }
+    }
+}
+
+/// LFU priority queue / set. Maintains up to `capacity` distinct keys; eviction
+/// removes the key with lowest frequency (ties broken by oldest insertion among
+/// that frequency bucket).
+pub struct LfuPQ<K> {
```

Comment:
nit: maybe the full name would be better. 
But it is a personal preference not a strong suggestion.

Reply:
Changed to long names.